### ISSUE TITLE
[QST] Any plan for being backpropagable

### ISSUE THREAD

**Issue by mpariente** [rapidsai/cusignal#64]
Thanks a lot for your work, this looks great ! 

Is there any plan to make some ops differentiable and backpropagable, maybe have a pytorch backend or something? 

Just asking because some of these things would have even a greater impact in deep learning based audio applications :-)

**Comment by awthomp** [comment]
Hey @mpariente -- thanks for participating and asking a question about cuSignal!

Currently, cuSignal's connection to PyTorch is via the `__cuda_array_interface__` which provides a standard for how GPU memory is structured and used. All libraries that support the interface can zero copy memory and keep data on the GPU as long as possible.

In the case of an audio signal processing operation, you could resample a waveform with `cusignal.resample_poly` and generate a spectrogram with `cusignal.spectrogram` or transition to the cepstrum domain with `cusignal.rceps` and then zero-copy pass that in-GPU memory to PyTorch to build your NN and backpropogate at will. In this vein, I imagine cuSignal is a pre-processing and collection library (both online and offline) that gets you all the way to the state of the data you're interested in before you start training.

If you're on PyTorch > 1.4, an example of this could be (Going from CuPy to PyTorch):

```python
# CuPy - GPU Array (like NumPy!)
gpu_arr = cp.random.rand(10_000, 10_000)

# Look at pointer
print('CuPy GPU Array Pointer: ', gpu_arr.__cuda_array_interface__['data'])

# Migrate from CuPy to PyTorch
torch_arr = torch.as_tensor(gpu_arr, device='cuda')

# Look at pointer -- it's the same as the CuPy array above!
print('PyTorch GPU Tensor Pointer: ', torch_arr.__cuda_array_interface__['data'])

# Migrate from PyTorch to CuPy
cupy_arr = cp.asarray(torch_arr)

# Look at pointer
print('CuPy GPU Pointer: ', cupy_arr.__cuda_array_interface__['data'])
```

In summary, if you have data on GPU that's been manipulated via a library that used `__cuda_array_interface__` (e.g. cuSignal, other RAPIDS libraries, CuPy, Numba, PyTorch, etc), you can pass pointers to access that data.

I hope this answers your questions. If you want the end-to-end pipeline to be differentiable and backpropagate-able, could you please give a bit more information about specific features you'd want? I tend to me more an RF guy and less a speech/audio guy, so even a list of functions that would be good for cuSignal inclusion would be really helpful!

**Comment by mpariente** [comment]
Hey @awthomp, thank you so much for your great answer, that is very informative !
I got what you explained about using CuSignal for online preprocessing and that's already great. I had something slightly different in mind. 

Let's say I have a waveform, I want to do some preprocessing, compute a spectrogram and then feed it to the network, the first steps can be done with CuSignal. 
Now, the output of my network is a modified spectrogram and I want to go back to the time domain to compute a time-domain loss function, I can't use CuSignal for that. And as spectrogram inversion is not universal, I'll probably end-up using a different spectrogram altogether. 

Another example : my loss function only works for a given sampling rate, I need to resample the signal before computing it, and then backprop, I can't do that with CuSignal. 

I hope this clarifies some of the use cases I had in mind.

**Comment by awthomp** [comment]
@mpariente 
>Let's say I have a waveform, I want to do some preprocessing, compute a spectrogram and then feed it to the network, the first steps can be done with CuSignal.
Now, the output of my network is a modified spectrogram and I want to go back to the time domain to compute a time-domain loss function, I can't use CuSignal for that. And as spectrogram inversion is not universal, I'll probably end-up using a different spectrogram altogether.

Are you imagining your classification to be some sort of segmentation task or classification? In other words, what objective are you training to? Do you know that X spectrogram makes up the phoneme \p\ or are you trying to detect where in the spectrogram something occurs?

>  Another example : my loss function only works for a given sampling rate, I need to resample the signal before computing it, and then backprop, I can't do that with CuSignal.

I'm also a bit confused here. Say you have a modulation recognition algorithm that's saying "this waveform is FM" and "this waveform is AM". When you're training the network, say you resample to rate X. I don't see why you'd have to resample further when training. If you want to classify different rates, can't you just have some waveforms at sample rate X and others at sample rate Y?

But to the broader point -- and I'm probably missing something cool you're trying to do, but when you're training these networks, you always have known input and known outputs -- so you'd know the pre-processing steps you used with cuSignal to get the data to that form. During inferencing, you can do the same pre-processing steps.

But you can also go from PyTorch back to cuSignal. The key part posted above:

```python
# For CuPy
cupy_arr = cp.asarray(torch_arr)

# For Numba
numba_arr_from_torch = cuda.to_device(torch_arr_numba)
```

Do you have any pseudo-code you can share so we can understand some of the cusignal limitations a bit more?

**Comment by mpariente** [comment]
> [...] You always have known input and known outputs -- so you'd know the pre-processing steps you used with cuSignal to get the data to that form. During inferencing, you can do the same pre-processing steps.

True for preprocessing. 

> Do you have any pseudo-code you can share so we can understand some of the cusignal limitations a bit more?

In general, I work in audio source separation and speech enhancement. Let's have a speech enhancement example in pytorch. 
```python
FS = 16000
def spectrogram(wav):
    return spec

def ispectrogram(spec):
    return wav

# Input mixture and target speech
noisy_speech = torch.randn(10, 1, 4*FS)
target_speech = torch.randn(10, 1, 4*FS)

# Compute spectrogram of noisy speech
noisy_spectrogram = spectrogram(noisy_speech)
# Get the estimated speech spectrogram
estimated_target_spectrogram = nnet(noisy_spectrogram)
# Get back to time domain
estimated_speech = ispectrogram(estimated_target_spectrogram)
# Compute time domain loss
loss = loss_func(estimated_speech, target_speech)
```

That was the first idea I had in mind. And I'd loose the gradients when I compute the ispectrogram with CuSignal right? 
Second thing I'd like to do : translate [this code](https://github.com/mpariente/pystoi) into pytorch to make it a backpropagable loss function. 
This metric works on 10kHz clean_speech, speech_estimate pairs. I don't want to reduce its usage only to 10kHz signal, I'd like to resample any input signal to 10kHz, in a backpropagable way.

**Comment by awthomp** [comment]
@mpariente This would be a neat thing to code up together and see where we can improve cusignal to meet your needs. I understand where you're coming from with backpropagation, but I *am* curious if we can essentially do this:

```python
# Compute spectrogram of noisy speech
noisy_spectrogram = spectrogram(noisy_speech)

# Get the estimated speech spectrogram
estimated_target_spectrogram = nnet(noisy_spectrogram)

# Move Torch array to cuSignal
estimated_target_spectrum = cupy.asarray(estimated_target_spectrogram)

# Get back to time domain - now in cuSignal
estimated_speech = ispectrogram(estimated_target_spectrogram)

# Move back to PyTorch
estimated_speech = torch.as_tensor(estimated_speech)

# Compute time domain loss
loss = loss_func(estimated_speech, target_speech)
```

I might be missing something with PyTorch details, but I don't immediately see why that wouldn't work. I assume there's a way to save or checkmark the gradients? Want to try together?

And thanks for pointing me to your projects. I'll dig in more this weekend, but I'd love to find places to work together!

Adam

**Comment by mpariente** [comment]
Hey Adam, 

> @mpariente This would be a neat thing to code up together and see where we can improve cusignal to meet your needs. I understand where you're coming from with backpropagation, but I _am_ curious if we can essentially do this:
> 
> ```python
> # Compute spectrogram of noisy speech
> noisy_spectrogram = spectrogram(noisy_speech)
> 
> # Get the estimated speech spectrogram
> estimated_target_spectrogram = nnet(noisy_spectrogram)
> 
> # Move Torch array to cuSignal
> estimated_target_spectrum = cupy.asarray(estimated_target_spectrogram)
> 
> # Get back to time domain - now in cuSignal
> estimated_speech = ispectrogram(estimated_target_spectrogram)
> 
> # Move back to PyTorch
> estimated_speech = torch.as_tensor(estimated_speech)
> 
> # Compute time domain loss
> loss = loss_func(estimated_speech, target_speech)
> ```
> 
> I might be missing something with PyTorch details, but I don't immediately see why that wouldn't work. I assume there's a way to save or checkmark the gradients? Want to try together?
>

This type of code would actually work, but if you call `loss.backward()`, things will break as `ispectrogram` doesn't have a `backward` method.
I'm pretty sure about that but not 100% so I'll make an example next week probably.

> And thanks for pointing me to your projects. I'll dig in more this weekend, but I'd love to find places to work together!


It would also be a pleasure! 
Manu
